### PR TITLE
crop or pad images using canvas size

### PR DESCRIPTION
**Pull request by chrisbank**
This is the initial implementation for canvas size cropping or padding. We will need some other tweaks but QE is satisfied this doesn't break any current functionality and is ready to go it.

### CODE REVIEW

File: lib/renderer.js [hunk 1]
```diff
@@ -719,34 +809,10 @@
...
-                        padding.left -= delta;
-                    }
-                    if (padding.top) {
-                        delta = Math.min(extract.y, padding.top);
-                        extract.y -= delta;
-                        extract.height += delta;
-                        padding.top -= delta;
-                    }
-                }
+                var canvasSizeScale = 1;
```
Comment (jaredadobe) [inline lib/renderer.js]: Add this to the var block above?

Comment (jaredadobe) [inline lib/renderer.js]: Are we losing scale here by having it set to 1?

File: lib/renderer.js [hunk 2]
```diff
@@ -552,6 +561,87 @@
...
+            if (dy > 0) {
+                padding.top += Math.floor(dy / 2);
+                padding.bottom += Math.ceil(dy / 2);
+            } else {
+                extract.y -= Math.floor(dy / 2);
+                extract.height = canvasScaledHeight;
+            }
+        }
+        
+        if ((padding.top || padding.left || padding.right || padding.bottom) ||
```
Comment (jaredadobe) [inline lib/renderer.js]: seems like object.keys().length > 0 would have the same effect?

File: lib/renderer.js [hunk 3]
```diff
@@ -552,6 +561,87 @@
...
+                                                                       convertSettings, canvasScale) {
+        var canvasScaledWidth = canvasWidth ? canvasWidth * canvasScale : 0,
+            canvasScaledHeight = canvasHeight ? canvasHeight * canvasScale : 0,
+            dx = canvasWidth ? canvasScaledWidth - pixmapWidth : 0,
+            dy = canvasHeight ? canvasScaledHeight - pixmapHeight : 0,
+            padding = convertSettings.padding || {top: 0, left: 0, right: 0, bottom: 0},
+            extract = convertSettings.extract || {x:0, y:0, width: pixmapWidth, height: pixmapHeight};
+        
+        if (dx) {
+            if (dx > 0) {
```
Comment (jaredadobe) [inline lib/renderer.js]: This check reads funny. It can never reach this line if it is 0.

File: lib/renderer.js (hunk 1 above)
Comment (chrisbank) [inline lib/renderer.js]: This is/was trying to figure out if canvas size should scale with image size. Decision for now was canvas is fixed and image size scales.

File: lib/renderer.js (hunk 3 above)
Comment (chrisbank) [inline lib/renderer.js]: done

File: lib/renderer.js (hunk 1 above)
Comment (chrisbank) [inline lib/renderer.js]: done

**Comment by jaredadobe** [comment]
LGTM when comments addressed. Merge when ready.

**Comment by chrisbank** [comment]
comments addressed. Will merge when travis goes green

**Pull request merged**